src/pso.c

```diff
@@ -0,0 +1,204 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <string.h>
+
+#include <time.h>
+#include <float.h>
+
+#include "pso.h"
+#include "utils.h"
+
+
+#define EPS 0.001
+#define COG 2
+#define SOCIAL 2
+#define INERTIA 0.5
+#ifndef M_PI
+#define M_PI (3.14159265358979323846)
+#endif
+
+#define min(a, b) (((a) < (b)) ? (a) : (b))
+#define max(a, b) (((a) > (b)) ? (a) : (b))
+
+  //returns swarm_size * dim size array
+double* rand_init(size_t swarm_size,
+                  size_t dim,
+                  const double* const min_positions,
+                  const double* const max_positions) {
+  double* init = (double*)malloc(swarm_size*dim*sizeof(double));
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      init[idx] = random_min_max(min_positions[d],max_positions[d]);
+    }
+  }
+  return init;
+}
+
+  //returns swarm_size size array
+void eval_fitness(double (*func)(const double* const, size_t),
+                  size_t swarm_size, size_t dim,
+                  const double* const x, double* fitness) {
+  for (size_t pop=0; pop<swarm_size; pop++){
+    fitness[pop] = func(x+(pop*dim),dim);
+  }
+  return ;
+}
+
+  //returns swarm_size x dim size array
+  // velocity at every posize_t of every particle
+double* gen_init_velocity(const double* const x,
+                          size_t swarm_size, size_t dim,
+                          const double* const min_positions,
+                          const double* const max_positions) {
+  double* u = (double *)malloc(swarm_size*dim*sizeof(double));
+  u = rand_init(swarm_size,dim,min_positions,max_positions);
+  double* v = (double *)malloc(swarm_size*dim*sizeof(double));
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      v[idx] = 0.25*(u[idx] - x[idx]);
+    }
+  }
+  return v;
+}
+
+  // sorts the population in decreasing order of fitness /
+  // i.e pop[0] posize_ts to best location yet.
+size_t best_fitness(double* fitness,size_t dim, size_t swarm_size) {
+  // find minimum of fitness
+  // set index of mimima as minima_idx;
+  // global_best = x+(dim*minima_idx);
+  size_t minima_idx = 0;
+  double best_yet = fitness[0]; // global best at timestep t+1
+  for (size_t p=1;p<swarm_size;p++){
+    if (fitness[p]<best_yet) {
+      best_yet = fitness[p];
+      minima_idx = p;
+    }
+  }
+  // global_best_position = local_best_position+(dim*minima_idx);
+  return minima_idx;
+}
+
+  // limit max and min velocity values to ensure no boundary overflow
+void generate_vel_limit(const double* min_positions,
+                        const double* max_positions,
+                        double* min_vel,
+                        double* max_vel,
+                        size_t dim) {
+  for (size_t d=0;d<dim;d++){
+    min_vel[d] = min_positions[d]/5;

```

Comment:
harcoded factor, define at top. is this factor given by the paper or is this kinda an arbitrary choice to scale the velocities wrt to position? Or make it a parameter of the function. But then still since it is a constant presumably it should be defined somewhere.

Reply:
yeah, this wasnt a part of the paper, I just decided to scale the min/max until i seemed to be getting a good enough value. It'll likely never change but I'll define it outside.